test/recipes/15-test_mp_rsa.t

```diff
@@ -0,0 +1,133 @@
+    # 3 primes, 2048-bit
+    {
+        primes => '3',
+        bits => '2048',
+    },
+    # 4 primes, 2048-bit
+    {
+        primes => '4',
+        bits => '2048',
+    },
+    # 8 primes, 2048-bit
+    {
+        primes => '8',
+        bits => '2048',
+    },
+    # 15 primes, 2048-bit
+    {
+        primes => '15',
+        bits => '2048',
+    },
+    # 8 primes, 15360-bit (3 & 4 primes for 15360 bit is too long to gen a key)
+    {
+        primes => '8',
+        bits => '15360',
+    },
+    # 15 primes, 15360-bit
+    {
+        primes => '15',
+        bits => '15360',
+    },
+);
+
+# genrsa
+run_mp_tests(0);
+# evp
+run_mp_tests(1);
+
+sub run_mp_tests {
+    my $evp = shift;
+
+    foreach my $param (@test_param) {
+        my $name = "evp$param->{bits}p$param->{primes}";
+        my $primes = $param->{primes};
+        my $bits = $param->{bits};
+
+        if (!$evp) {
+            $name = "$param->{bits}p$param->{primes}";
+        }
+
+        if ($evp) {
+            ok(run(app([ 'openssl', 'genpkey', '-out', 'rsamptest.pem',
+                         '-algorithm', 'RSA', '-pkeyopt', "rsa_keygen_primes:$primes",
+                         '-pkeyopt', "rsa_keygen_bits:$bits"])), "genrsa $name");
+        } else {
+            ok(run(app([ 'openssl', 'genrsa', '-out', 'rsamptest.pem',
+                         '-primes', $primes, $bits])), "genrsa $name");
+        }
+
+        ok(run(app([ 'openssl', 'rsa', '-check', '-in', 'rsamptest.pem',
+                     '-noout'])), "rsa -check $name");
+        if ($evp) {
+            ok(run(app([ 'openssl', 'pkeyutl', '-inkey', 'rsamptest.pem',
+                         '-encrypt', '-in', "$cleartext",
+                         '-out', 'rsamptest.enc' ])), "rsa $name encrypt");
+            ok(run(app([ 'openssl', 'pkeyutl', '-inkey', 'rsamptest.pem',
+                         '-decrypt', '-in', 'rsamptest.enc',
+                         '-out', 'rsamptest.dec' ])), "rsa $name decrypt");
+        } else {
+            ok(run(app([ 'openssl', 'rsautl', '-inkey', 'rsamptest.pem',
+                         '-encrypt', '-in', "$cleartext",
+                         '-out', 'rsamptest.enc' ])), "rsa $name encrypt");
+            ok(run(app([ 'openssl', 'rsautl', '-inkey', 'rsamptest.pem',
+                         '-decrypt', '-in', 'rsamptest.enc',
+                         '-out', 'rsamptest.dec' ])), "rsa $name decrypt");
+        }
+
+        ok(check_msg(), "rsa $name check result");
+
+        # clean up temp files
+        unlink 'rsamptest.pem';
+        unlink 'rsamptest.enc';
+        unlink 'rsamptest.dec';
+    }
+}
+
+sub check_msg {
+    open(my $fh, "<", "$cleartext") or return 0;
+    my @msg = <$fh>;
+    close $fh;
+    open($fh, "<", "rsamptest.dec") or return 0;
+    my @dec = <$fh>;
+    close $fh;
+
+    @msg = map { s|\R$||; $_ } (@msg);
+    @dec = map { s|\R$||; $_ } (@dec);
+
+    if (scalar(@msg) != scalar(@dec)) {
+        print "length not the same";
```

Comment:
The message is misleading, because it's amount of *lines* that is compared, not length of the message.

Reply:
I don't see any reason for why test message can't be a temporary file. And if it's always meant to be single line, or more specifically no new lines, then one could simplify verification to scalar(@dec) != 1 followed by single string comparison.

Reply:
Hmm, it can't guarantee that the message will be a single line, it might be updated in the future, so the current `check_msg` function is more versatile, kind of 'once and for all'...